Code/Tools/SceneAPI/SceneCore/Tests/Containers/Views/PairIteratorTests.cpp

```diff
@@ -19,6 +19,18 @@
 #include <SceneAPI/SceneCore/Containers/Views/PairIterator.h>
 #include <SceneAPI/SceneCore/Tests/Containers/Views/IteratorTestsBase.h>
 
+// This test gives trouble with /permissive-, the following instantiation workarounds the missing resolution
+namespace std
+{
+    template<>
+    void iter_swap(
+        AZ::SceneAPI::Containers::Views::PairIterator<int*, int*, std::random_access_iterator_tag> lhs,
+        AZ::SceneAPI::Containers::Views::PairIterator<int*, int*, std::random_access_iterator_tag> rhs)
+    {
+        AZStd::iter_swap(lhs, rhs);
+    }
+}
+
```

Comment:
This seems like it might cause ODR violations, even if it does compile. What is calling `std::iter_swap` in these tests that is causing problems? Should this instead be defined in the `AZ::SceneAPI::Containers::Views` namespace (in `Code/Tools/SceneAPI/SceneCore/Containers/Views/PairIterator.h`) and let ADL do its thing?

Reply:
the different algorithms are the ones calling it (e.g. `AZStd::reverse(view.begin(), view.end());`)
For some reason, its not picking up the definition at the end of `Code\Tools\SceneAPI\SceneCore\Containers\Views\PairIterator.inl`
I think is due to that 3rd parameter (the iterator tag), but we tried different combinations of things with @lumberyard-employee-dm and couldnt figure it out.

Reply:
`PairIterator.inl` defines a free function template called `iter_swap` in the `AZStd` namespace, for 2 `PairIterator` iterators.
`AZStd::reverse` is an alias for `std::reverse`.
`std::reverse` is (in libstdc++, at least) implemented by calling `std::iter_swap`.
`std::iter_swap` is a function template that accepts two different types of iterators. It is implemented by calling `swap(*iter1, *iter2)`.
At this point, because the call to `swap` is not qualified with a namespace, ADL kicks in, and the `swap` function that is called depends on the type of the arguments, and the namespace that those arguments is defined in.
For `PairIterator`, its `operator*()` method returns a "`reference`":
```
template<typename FirstIterator, typename SecondIterator>
class PairIterator<FirstIterator, SecondIterator, void>
{
public:
    using reference = AZStd::pair<typename AZStd::iterator_traits<FirstIterator>::reference, typename AZStd::iterator_traits<SecondIterator>::reference>;
};
                                                                                                                                                             
// Input iterator
template<typename FirstIterator, typename SecondIterator>
class PairIterator<FirstIterator, SecondIterator, AZStd::input_iterator_tag>
    : public PairIterator<FirstIterator, SecondIterator, void>
{
public:
    using super = PairIterator<FirstIterator, SecondIterator, void>;
                                                                                                                                                             
    typename super::reference operator*() const
    {
        return typename super::reference(*super::m_first, *super::m_second);
    }
};
```

Note that `reference` type: it is *not a reference*, but a value, namely a `AZStd::pair<T1&, T2&>`.

ADL sees that type is in the `AZStd` namespace, so tries to match that argument to the `AZStd::swap` function for pairs, but that function needs l-values:
```
/ws/main/Code/Framework/AzCore/./AzCore/std/utils.h:245:20: note: candidate function [with T1 = int &, T2 = int &] not viable: expects an l-value for 1st argument
    constexpr void swap(AZStd::pair<T1, T2>& left, AZStd::pair<T1, T2>& _Right)
```

I would guess that this started failing due to removing the AZStd version of `reverse` and swapping it for an alias to the `std` implementation.

I think the best thing to do here is to update the `iter_swap` function template specialization in `PairIterator.h` to be in the `std` namespace instead.

Reply:
I think the benefit of getting this change in outweighs any specific problems with this test.